core/server/master/src/main/java/alluxio/master/file/AccessTimeUpdater.java

```diff
@@ -0,0 +1,176 @@
+
+import java.util.Iterator;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * This class represents the executor for periodic inode ttl check.
+ */
+@NotThreadSafe
+final class AccessTimeUpdater implements JournalSink {
+  private static final Logger LOG = LoggerFactory.getLogger(AccessTimeUpdater.class);
+
+  private final long mFlushInterval;
+  private final long mUpdatePrecision;
+  private final long mShutdownTimeout;
+
+  private final FileSystemMaster mFileSystemMaster;
+  private final InodeTree mInodeTree;
+
+  /** Keep track of all inodes that need access time update. */
+  private ConcurrentHashSet<Long> mAccessTimeUpdates;
+  private ScheduledExecutorService mExecutorService = null;
+  private AtomicBoolean mUpdateScheduled = new AtomicBoolean();
+
+  /**
+   * Constructs a new {@link AccessTimeUpdater}.
+   */
+  public AccessTimeUpdater(FileSystemMaster fileSystemMaster, InodeTree inodeTree,
+      JournalSystem journalSystem) {
+    this(fileSystemMaster, inodeTree, journalSystem,
+        ServerConfiguration.getMs(PropertyKey.MASTER_FILE_ACCESS_TIME_JOURNAL_FLUSH_INTERVAL),
+        ServerConfiguration.getMs(PropertyKey.MASTER_FILE_ACCESS_TIME_UPDATE_PRECISION),
+        ServerConfiguration.getMs(PropertyKey.MASTER_FILE_ACCESS_TIME_UPDATER_SHUTDOWN_TIMEOUT));
+  }
+
+  /**
+   * Constructs a new {@link AccessTimeUpdater} with time configurations.
+   */
+  @VisibleForTesting
+  public AccessTimeUpdater(FileSystemMaster fileSystemMaster, InodeTree inodeTree,
+      JournalSystem journalSystem, long flushInterval, long updatePrecision, long shutdownTimeout) {
+    mFileSystemMaster = fileSystemMaster;
+    mInodeTree = inodeTree;
+    mAccessTimeUpdates = new ConcurrentHashSet<>();
+    mFlushInterval = flushInterval;
+    mUpdatePrecision = updatePrecision;
+    mShutdownTimeout = shutdownTimeout;
+    journalSystem.addJournalSink(mFileSystemMaster, this);
+  }
+
+  public void start() {
+    start(mFlushInterval > 0 ? Executors.newSingleThreadScheduledExecutor(
+        ThreadFactoryUtils.build("AccessTimeUpdater-%d", true)) : null);
+  }
+
+  @VisibleForTesting
+  public void start(ScheduledExecutorService executorService) {
+    mExecutorService = executorService;
+  }
+
+  @Override
+  public void beforeShutdown() {
+    if (mExecutorService != null) {
+      flushUpdates();
+    }
+  }
+
+  public void stop() {
+    if (mExecutorService != null) {
+      ThreadUtils.shutdownAndAwaitTermination(mExecutorService, mShutdownTimeout);
+    }
+  }
+
+  /**
+   * Update inode last access time.
+   *
+   * @param context the journal context
+   * @param inode the target inode
+   * @param opTimeMs the last access time to set on the inode
+   */
+  public void updateAccessTime(JournalContext context, Inode inode, long opTimeMs) {
```

Comment:
is this method run while already holding the lock on the inode, in FSM?

Reply:
Yes. I will add a javadoc comment to note that.